stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+                !Self::have_unprocessed_tenures(
+                    sortdb
+                        .pox_constants
+                        .block_height_to_reward_cycle(
+                            sortdb.first_block_height,
+                            self.nakamoto_start_height,
+                        )
+                        .expect("FATAL: first nakamoto block from before system start"),
+                    &self.tenure_downloads.completed_tenures,
+                    prev_wanted_tenures,
+                    &self.tenure_block_ids,
+                    &sortdb.pox_constants,
+                    sortdb.first_block_height,
+                    invs.inventories.values(),
+                )
+            } else {
+                test_debug!("No prev_wanted_tenures yet");
+                true
+            };
+
+        if can_advance_wanted_tenures && self.reward_cycle != sort_rc {
+            let mut prev_wanted_tenures = vec![];
+            let mut cur_wanted_tenures = vec![];
+            let prev_wts = self.prev_wanted_tenures.take().unwrap_or(vec![]);
+            let cur_wts = std::mem::replace(&mut self.wanted_tenures, vec![]);
+
+            for wt in new_wanted_tenures
+                .into_iter()
+                .chain(prev_wts.into_iter())
+                .chain(cur_wts.into_iter())
+            {
+                test_debug!("Consider wanted tenure: {:?}", &wt);
+                let wt_rc = sortdb
+                    .pox_constants
+                    .block_height_to_reward_cycle(sortdb.first_block_height, wt.burn_height)
+                    .expect("FATAL: height before system start");
+                if wt_rc + 1 == sort_rc {
+                    prev_wanted_tenures.push(wt);
+                } else if wt_rc == sort_rc {
+                    cur_wanted_tenures.push(wt);
+                } else {
+                    test_debug!("Drop wanted tenure: {:?}", &wt);
+                }
+            }
+
+            prev_wanted_tenures.sort_by(|wt1, wt2| wt1.burn_height.cmp(&wt2.burn_height));
+            cur_wanted_tenures.sort_by(|wt1, wt2| wt1.burn_height.cmp(&wt2.burn_height));
+
+            test_debug!("prev_wanted_tenures is now {:?}", &prev_wanted_tenures);
+            test_debug!("wanted_tenures is now {:?}", &cur_wanted_tenures);
+
+            self.prev_wanted_tenures = if prev_wanted_tenures.len() > 0 {
+                Some(prev_wanted_tenures)
+            } else {
+                None
+            };
+            self.wanted_tenures = cur_wanted_tenures;
+            self.reward_cycle = sort_rc;
+        } else {
+            test_debug!(
+                "Append {} wanted tenures: {:?}",
+                new_wanted_tenures.len(),
+                &new_wanted_tenures
+            );
+            self.wanted_tenures.append(&mut new_wanted_tenures);
+            test_debug!("wanted_tenures is now {:?}", &self.wanted_tenures);
+        }
```

Comment:
Slight preference to reverse this if statement-- I find it a lot easier to read `else` cases if the original `if` is still in view, and so this "easier" case can go first. But looking at this `else`, it seems off to me:

```rust
if can_advance_wanted_tenures && self.reward_cycle != sort_rc {
    ...
} else {
    // this case is evaluated if `can_advance_wanted_tenures = false`. 
}    
```

Maybe I'm just misreading the meaning of `can_advance_wanted_tenures` -- it seems fine to append to `cur_wanted_tenures` even if there's still pending `prev_wanted_tenures`.

Reply:
Yes -- that happens if you're at the chain tip and the burnchain advances by one block.  `prev_wanted_tenures` holds the last reward cycle's tenures, and `wanted_tenures` holds the current reward cycle's tenures (which will naturally grow until we hit a reward cycle boundary, at which point `can_process_unwanted_tenures` will be `true` if all `wanted_tenures` are downloaded and `sort_rc != self.reward_cycle` will be `true` as well).